lantiq: add support for AVM FRITZ!Box 7362 SL

danijelt opened the pull request:
[1/2] lantiq: add support for AVM FRITZ!Box 7362 SL
Hardware:
- SoC: Lantiq VRX288
- RAM: Winbond W971GG6JB 128 MiB
- Flash:
  - SPI: 8 Mb (1 MiB) for bootloader and tffs
  - NAND: 1Gb (128 MiB) for OS
- xDSL: Lantiq VRX208
- WLAN: Atheros AR9381
- DECT: Dialog Semiconductors SC14441

Everything except FXS/DECT works
(no drivers for AVM's FXS implementation with SC14441).

Installation via FTP:
1. Use scripts/flashing/eva_ramboot.py to send initramfs-kernel.bin
to the device when powering on.
Standard AVM procedures with finding the correct IP address and
the right moment to open FTP apply here (approx. 4 seconds on 7362SL).
IMPORTANT: set lzma compression in ramdisk options, bootloader stalls
when receiving uncompressed images.

2. Transfer sysupgrade.bin image with scp to /tmp directory
and run sysupgrade

3. First boot might take a bit longer if linux_fs_start was set to 1,
in that case the device will reboot twice, first time it will fail to load
second kernel (overwritten by ubifs), set linux_fs_start to 0 and reboot.

OpenWrt uses the entire NAND flash. Kernel uses 4 MiB and rootfs uses
the rest of 124 MiB, overwriting everything related to FRITZ!OS - both
OS images, config and answering machine/media server data.

[2/2] lantiq: unify FRITZ736X device trees
FRITZ!Box 7360 and FRITZ!Box 7362 share similar configuration,
unify it in FRITZ736X.dtsi.

ynezz commented on target/linux/lantiq/files-4.14/arch/mips/boot/dts/FRITZ7362SL.dts:
```diff
@@ -0,0 +1,285 @@
+/dts-v1/;
```
@danijelt  Please can you consider `SPDX-License-Identifier: GPL-2.0-or-later OR MIT` as in [DTS Checklist](https://openwrt.org/submitting-patches#dts_checklist) ?
> Having MIT as alternative is sometimes preferred by upstream maintainers
and allows sharing that simple code with other projects. We don't really
want multiple DTS versions for the same device.

danijelt replied:
Done.

ynezz commented on target/linux/lantiq/files-4.14/arch/mips/boot/dts/FRITZ7362SL.dts:
```diff
@@ -0,0 +1,285 @@
+		reg = <0x0 0x8000000>;
+	};
+
+	gpio-keys-polled {
```
@danijelt  Could you please change this to generic name `keys` as suggested by [DTS Checklist](https://openwrt.org/submitting-patches#dts_checklist)
>  5. The name of a node should reflect the function of the device and not its model. Examples for generic node names can be found in Section 2.2.2 Generic Names Recommendation

danijelt replied:
Done.

ynezz commented on target/linux/lantiq/files-4.14/arch/mips/boot/dts/FRITZ7362SL.dts:
```diff
@@ -0,0 +1,285 @@
+		};
+	};
+
+	gpio-leds {
```
@danijelt  Could you please change this to generic name `leds` as suggested by [DTS Checklist](https://openwrt.org/submitting-patches#dts_checklist)
>  5. The name of a node should reflect the function of the device and not its model. Examples for generic node names can be found in Section 2.2.2 Generic Names Recommendation

danijelt replied:
Done.

ynezz commented on target/linux/lantiq/files-4.14/arch/mips/boot/dts/FRITZ7362SL.dts:
```diff
@@ -0,0 +1,285 @@
+	pinctrl-names = "default";
+	pinctrl-0 = <&pins_spi_default>;
+
+	m25l80@4 {
```
@danijelt Can you please try to use generic name `flash@4` and see if it boots with this change?

danijelt replied:
Done. Device boots.

ynezz commented on target/linux/lantiq/image/Makefile:
```diff
@@ -664,6 +664,18 @@ define Device/avm_fritz7360sl
+  DEVICE_DTS := FRITZ7362SL
+  DEVICE_TITLE := AVM FRITZ!Box 7362 SL
+  DEVICE_PACKAGES := kmod-ath9k kmod-owl-loader wpad-basic kmod-usb-dwc2 fritz-tffs
+  SUPPORTED_DEVICES += FRITZ7362SL
```
@danijelt You can drop this line, since you're adding support for the new device which is added automatically for you in

https://github.com/openwrt/openwrt/blob/75fdf3ba01e9379f4301ccbaf3eeedf0923076f1/target/linux/lantiq/image/Makefile#L64

danijelt replied:
Done.

ynezz commented:
@danijelt You've a lot of common with FRITZ7360SL.dts so you need to create DTSI include file FRITZ736xSL.dtsi and add there common nodes for 7360 and 7362 and then include this DTSI in each DTS file.

While at it, try to make formatting of the DTS file consistent, try to separate every DTS node with newline like you've it for example in `m25l80@4` node. Do the same for rest of the nodes.

danijelt commented:
> @danijelt You've a lot of common with FRITZ7360SL.dts so you need to create DTSI include file FRITZ736xSL.dtsi and add there common nodes for 7360 and 7362 and then include this DTSI in each DTS file.
> 
> While at it, try to make formatting of the DTS file consistent, try to separate every DTS node with newline like you've it for example in `m25l80@4` node. Do the same for rest of the nodes.

Should I unify the LED labels under "fritz736xsl:*:*"? Their definitions and functions are the same.

ynezz commented:
> Should I unify the LED labels under "fritz736xsl::"? Their definitions and functions are the same.

I don't think so, since you would break sysupgrade on existing `fritz7360` devices.

LipkeGu commented:
I have the 7360 SL so i can support you with testing

danijelt commented:
@ynezz In the meantime, 7360 naming was fixed upstream, so I rebased my repo and eliminated that first commit that fixed it.

danijelt commented:
@ynezz Any news?

ynezz commented on target/linux/lantiq/files-4.14/arch/mips/boot/dts/FRITZ7362SL.dts:
```diff
@@ -0,0 +1,286 @@
+
+	keys {
+		compatible = "gpio-keys-polled";
+		#address-cells = <1>;
```
@danijelt You can drop `#address-cells = <1>;` and `#size-cells = <0>;` as it's not necessary.

danijelt replied:
Done.

ynezz commented:
@danijelt It's getting better, thanks!

BTW, I've [tried suggest to you](https://github.com/openwrt/openwrt/pull/1742#issuecomment-453785118), that it would be nice, if you could try to formatting of the DTS files consistent, try to separate every DTS node with newline like you've it for example in `flash@4` node. Do the same for rest of the nodes.

@blocktrron Could you please take a look as well?

ynezz commented:
@danijelt I've just downloaded this PR locally to build test it, and I've found out, that you've done the refactor in the wrong order.

* `lantiq: unify FRITZ736X device trees` should be first commit `lantiq: Move common FRITZ736X bits into include file`
* after that you can add support for `7362 SL` device

Not the other way around like you've it now, where you've added support for `7362 SL` and then did the refactor. That's wrong.

ynezz commented:
@danijelt For example I can't find following fragment from local `FRITZ7362SL.dts` in GitHub interface:

```
&pcie0 {
        gpio-reset = <&gpio 21 GPIO_ACTIVE_LOW>;

        pcie@0 {
                #size-cells = <1>;
                #address-cells = <2>;
        };
};
```

Commit:
lantiq: move common FRITZ736X bits into dtsi

Move common FRITZ736X bits into dtsi file
as a preparation for FRITZ7362SL support.

Signed-off-by: Danijel Tudek <danijel.tudek@gmail.com>

Commit:
lantiq: add support for AVM FRITZ!Box 7362 SL

Hardware:
- SoC: Lantiq VRX288
- RAM: Winbond W971GG6JB 1 Gb (128 MiB)
- Flash:
  - SPI: 8 Mb (1 MiB) for bootloader and tffs
  - NAND: 1 Gb (128 MiB) for OS
- xDSL: Lantiq VRX208
- WLAN: Atheros AR9381
- DECT: Dialog Semiconductors SC14441

Everything except FXS/DECT works
(no drivers for AVM's FXS implementation with SC14441).

Installation via FTP:
1. Use scripts/flashing/eva_ramboot.py to send initramfs-kernel.bin
to the device when powering on.
Standard AVM procedures with finding the correct IP address and
the right moment to open FTP apply here (approx. 4 seconds on 7362SL).
IMPORTANT: set lzma compression in ramdisk options, bootloader stalls
when receiving uncompressed images.

2. Transfer sysupgrade.bin image with scp to /tmp directory
and run sysupgrade

3. First boot might take a bit longer if linux_fs_start was set to 1,
in that case the device will reboot twice, first time it will fail to load
second kernel (overwritten by ubifs), set linux_fs_start to 0 and reboot.

OpenWrt uses the entire NAND flash. Kernel uses 4 MiB and rootfs uses
the rest of 124 MiB, overwriting everything related to FRITZ!OS - both
OS images, config and answering machine/media server data.
To return to FRITZ!OS, use AVM's recovery image.

Signed-off-by: Danijel Tudek <danijel.tudek@gmail.com>

danijelt commented:
@ynezz Can you check it now? I've reordered it and fixed style as suggested.

danijelt commented:
@ynezz Sorry, I noticed a typo and fixed it, everything should be OK now.

hauke commented:
Thank you for the patch it was applied to master.
